modules/internal/DefaultAssociative.chpl

```diff
@@ -106,16 +106,81 @@ module DefaultAssociative {
     }
   
     proc dsiSerialReadWrite(f /*: Reader or Writer*/) {
-      var first = true;
-      f <~> new ioLiteral("{");
-      for idx in this {
-        if first then 
-          first = false; 
-        else 
-          f <~> new ioLiteral(", ");
-        f <~> idx;
-      }
-      f <~> new ioLiteral("}");
+
+      var binary = f.binary();
+
+      if f.writing {
+        if binary {
+          var numIndices: int = dsiNumIndices;
+          f <~> numIndices;
+          for idx in this {
+            f <~> idx;
+          }
+        } else {
+          var first = true;
+          f <~> new ioLiteral("{");
+          for idx in this {
+            if first then 
+              first = false; 
+            else 
+              f <~> new ioLiteral(", ");
+            f <~> idx;
+          }
+          f <~> new ioLiteral("}");
+        }
+      } else {
+        if binary {
+          var numIndices: int;
```

Comment:
Should this clear the domain first to make sure it only contains the indices that are read in?